x-pack/plugin/search-business-rules/src/main/java/org/apache/lucene/search/CappedScoreQuery.java

```diff
@@ -0,0 +1,170 @@
+                        @Override
+                        public float score() throws IOException {
+                            return Math.min(maxScore, in.score());
+                        }
+                    });
+                }
+            };
+        }
+
+        @Override
+        public long cost() {
+            return bulkScorer.cost();
+        }
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        final Weight innerWeight = searcher.createWeight(query, ScoreMode.COMPLETE, 1f);
+        if (scoreMode.needsScores()) {
+            return new CappedScoreWeight(this, innerWeight, boost) {
+                @Override
+                public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
+                    if (scoreMode == ScoreMode.TOP_SCORES) {
+                        return super.bulkScorer(context);
+                    }
+                    final BulkScorer innerScorer = innerWeight.bulkScorer(context);
+                    if (innerScorer == null) {
+                        return null;
+                    }
+                    return new CappedBulkScorer(innerScorer, this, maxScore);
+                }
+
+                @Override
+                public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOException {
+                    ScorerSupplier innerScorerSupplier = innerWeight.scorerSupplier(context);
+                    if (innerScorerSupplier == null) {
+                        return null;
+                    }
+                    return new ScorerSupplier() {
+                        @Override
+                        public Scorer get(long leadCost) throws IOException {
+                            final Scorer innerScorer = innerScorerSupplier.get(leadCost);
+                            return new CappedScorer(innerWeight, innerScorer, maxScore);
+                        }
+
+                        @Override
+                        public long cost() {
+                            return innerScorerSupplier.cost();
+                        }
+                    };
+                }
+
+                @Override
+                public Matches matches(LeafReaderContext context, int doc) throws IOException {
+                    return innerWeight.matches(context, doc);
+                }
+
+                @Override
+                public Scorer scorer(LeafReaderContext context) throws IOException {
+                    ScorerSupplier scorerSupplier = scorerSupplier(context);
+                    if (scorerSupplier == null) {
+                        return null;
+                    }
+                    return scorerSupplier.get(Long.MAX_VALUE);
+                }
+
+                @Override
+                public boolean isCacheable(LeafReaderContext ctx) {
+                    return innerWeight.isCacheable(ctx);
+                }
```

Comment:
should we move this logic to CappedScoreWeight directly instead of having this in a sub class?